exec/java-exec/src/main/java/org/apache/drill/exec/store/StoragePluginRegistryImpl.java

```diff
@@ -36,608 +30,616 @@
-      Map<String, URL> pluginURLMap = new HashMap<>();
-      for (URL url : storageUrls) {
-        loadStoragePlugins(url, bootstrapPlugins, pluginURLMap, lpPersistence);
+  private void upgradeStore() {
+    StoragePlugins upgraded = new StoragePlugins();
+    for (ConnectorLocator locator : locators) {
+      StoragePlugins locatorPlugins = locator.updatedPlugins();
+      if (upgraded != null) {
+        upgraded.putAll(locatorPlugins);
       }
-      if (formatUrls != null && !formatUrls.isEmpty()) {
-        logger.info("Loading the format plugin configs from URLs {}.", formatUrls);
-        for (URL url : formatUrls) {
-          loadFormatPlugins(url, bootstrapPlugins, pluginURLMap, lpPersistence);
-        }
+    }
+    if (upgraded.isEmpty()) {
+      return;
+    }
+    for (Map.Entry<String, StoragePluginConfig> newPlugin : upgraded) {
+      StoragePluginConfig oldPluginConfig = pluginStore.get(newPlugin.getKey());
+      if (oldPluginConfig != null) {
+        copyPluginStatus(oldPluginConfig, newPlugin.getValue());
       }
-      return bootstrapPlugins;
-    } else {
-      throw new IOException("Failure finding " + ExecConstants.BOOTSTRAP_STORAGE_PLUGINS_FILE);
+      pluginStore.put(newPlugin.getKey(), newPlugin.getValue());
     }
   }
 
   /**
-   * Loads storage plugins from the given URL
+   * Identifies the enabled status for new storage plugins
+   * config. If this status is absent in the updater file, the status is kept
+   * from the configs, which are going to be updated
    *
-   * @param url URL to the storage plugins bootstrap file
-   * @param bootstrapPlugins a collection where the plugins should be loaded to
-   * @param pluginURLMap a map to store correspondence between storage plugins and bootstrap files in which they are defined. Used for logging
-   * @param lpPersistence need to get an object mapper for the bootstrap files
-   * @throws IOException if failed to retrieve a plugin from a bootstrap file
+   * @param oldPluginConfig
+   *          current storage plugin config from Persistent Store or bootstrap
+   *          config file
+   * @param newPluginConfig
+   *          new storage plugin config
    */
-  private void loadStoragePlugins(URL url, StoragePlugins bootstrapPlugins, Map<String, URL> pluginURLMap, LogicalPlanPersistence lpPersistence) throws IOException {
-    StoragePlugins plugins = getPluginsFromResource(url, lpPersistence);
-    plugins.forEach(plugin -> {
-      StoragePluginConfig oldPluginConfig = bootstrapPlugins.putIfAbsent(plugin.getKey(), plugin.getValue());
-      if (oldPluginConfig != null) {
-        logger.warn("Duplicate plugin instance '[{}]' defined in [{}, {}], ignoring the later one.",
-            plugin.getKey(), pluginURLMap.get(plugin.getKey()), url);
-      } else {
-        pluginURLMap.put(plugin.getKey(), url);
-      }
-    });
+  protected static void copyPluginStatus(
+      @NotNull StoragePluginConfig oldPluginConfig,
```

Comment:
Please remove @NotNull annotation, it's from javax.annotations and it's a separate lib which can access transitively.